src/main/java/com/digitalasset/testing/cucumber/steps/LedgerInteractions.java

```diff
@@ -0,0 +1,214 @@
+                        party(party), idWithArgs.identifier, contractId, choiceName, args);
+              }
+            });
+    Then(
+        "^.*\"([^\"]+)\" should observe the creation of \"([^\"]+)\"$",
+        (String party, String moduleAndEntityName) -> {
+          PackageUtils.TemplateType idWithArgs =
+              findTemplate(sandboxCommunicator.getClient(), moduleAndEntityName);
+          sandboxCommunicator
+              .getLedgerAdapter()
+              .getCreatedContractId(party(party), idWithArgs.identifier, ContractId::new);
+        });
+    Then(
+        "^.*\"([^\"]+)\" should observe the creation of \"([^\"]+)\" with(?: contract id \"([^\"]+)\" and)? values$",
+        (String party, String moduleAndEntityName, String contractId, DataTable dataTable) -> {
+          PackageUtils.TemplateType idWithArgs =
+              findTemplate(sandboxCommunicator.getClient(), moduleAndEntityName);
+          Record args = fieldsToArgs(checkTableIsList(dataTable), idWithArgs.createFields);
+          sandboxCommunicator
+              .getLedgerAdapter()
+              .observeEvent(
+                  party,
+                  ContractCreated.expectContractWithArguments(
+                      idWithArgs.identifier, "{CAPTURE:" + contractId + "}", args));
+        });
+    Then(
+        "^.*\"([^\"]+)\" should observe the archival of \"([^\"]+)\" with contract id \"([^\"]+)\".*$",
+        (String party, String moduleAndEntityName, String contractIdKey) -> {
+          PackageUtils.TemplateType idWithArgs =
+              findTemplate(sandboxCommunicator.getClient(), moduleAndEntityName);
+          ContractId contractId =
+              sandboxCommunicator
+                  .getLedgerAdapter()
+                  .valueStore
+                  .get(contractIdKey)
+                  .asContractId()
+                  .get();
+          sandboxCommunicator
+              .getLedgerAdapter()
+              .observeEvent(
+                  party, ContractArchived.apply(idWithArgs.identifier.toString(), contractId));
+        });
+    Then(
+        "^.*they should receive a technical failure (with|containing) message \\s*\"([^\"]*)\".*$",
+        (String withOrContaining, String errorMessageRegex) -> {
+          Throwable lastResult = resultHolder.getAndSet(null);
+          if (lastResult == null) {
+            throw new AssertionError("Failure was expected but not observed");
+          }
+          if (!withOrContaining.equals(WITH)) {
+            errorMessageRegex = ".*" + errorMessageRegex + ".*";
+          }
+          assertTrue(
+              "Expected error text [" + errorMessageRegex + "], but found [" + lastResult + "]",
+              Pattern.compile(errorMessageRegex, Pattern.DOTALL)
+                  .matcher(lastResult.toString())
+                  .matches());
+        });
+  }
+
+  private List<String> checkTableIsList(DataTable dataTable) {
+    if (dataTable.height() == 1 & dataTable.width() > 0) {
```

Comment:
Minor comment, you can ignore: I suggest `&&` instead of `&`.

Reply:
Ouch, thanks for noticing.